Project1/Map.cpp

```diff
@@ -0,0 +1,135 @@
+	if (destination->getTile().hasUnit()) {
+		return false;
+	}
+	else {
+		_unit_to_cell.insert(std::pair<Unit*, GridCell*>(new_unit, destination));// Map the Unit to the destination.
+		destination->getTile().insertUnit(new_unit);
+		return true;
+	}
+}
+void Map::removeUnit(Unit* unit) { //throws exceptions if the space is not empty.
+	_unit_to_cell.at(unit)->getTile().removeUnit();
+	_unit_to_cell.erase(unit);
+}
+
+std::vector<GridCell*> Map::getAccesibleCells(Unit* unit) {
+	findShortestPaths(_unit_to_cell.at(unit), unit->getStats().getAttribs()[AttribType::values::MOV], unit->getStats().getMobilityType());
+	std::vector<GridCell*> cells = std::vector<GridCell*>();
+	for (std::map<GridCell*, prev_cost_pair>::iterator it = _shortest_path_map.begin(); it != _shortest_path_map.end(); it++) {
+		cells.push_back(it->first);
+	}
+	return cells;
+}
+CellPath Map::getShortestPath(GridCell* start, GridCell* destination) {
+	findShortestPaths(start);
+	if (start == destination) {
+		return CellPath(start);
+	}
+	CellPath path = getShortestPath(start, _shortest_path_map.at(destination)._prev_cell);
+	path.insertTile(destination);
+	return path;
+}
+void Map::findShortestPaths(GridCell* start) {
+	if (start == _shortest_path_start) { //need another flag for if the map is changed
+		return;
+	}
+	std::priority_queue<prev_cost_pair> queue = std::priority_queue<prev_cost_pair>();
+	prev_cost_pair first;
+	first._cost = 0;
+	first._prev_cell = start;
+	_shortest_path_map.clear();
+	_shortest_path_map.insert(std::pair<GridCell*, prev_cost_pair>(start, first));
+	queue.push(first);
+	findShortestPaths(queue, -1, _MOBILITY_TYPES::GROUNDED);//assume grounded for now
+}
+void Map::findShortestPaths(GridCell* start, int max_move, MobilityType mobility) {
+	if (start == _shortest_path_start) { //need another flag for if the map is changed
+		return;
+	}
+	std::priority_queue<prev_cost_pair> queue = std::priority_queue<prev_cost_pair>();
+	prev_cost_pair first;
+	first._cost = 0;
+	first._prev_cell = start;
+	_shortest_path_map.clear();
+	_shortest_path_map.insert(std::pair<GridCell*, prev_cost_pair>(start, first));
+	queue.push(first);
+	findShortestPaths(queue, max_move, mobility);
+}
+void Map::findShortestPaths(std::priority_queue<prev_cost_pair>& queue, int max_move, MobilityType mobility) {
+	if (queue.size() == 0) {
+		return;
+	}
+	prev_cost_pair top = queue.top();
+	queue.pop();
+	std::vector<GridCell*> adj_cells = top._prev_cell->getAdjacentCells();
+	for (std::vector<GridCell*>::iterator it = adj_cells.begin(); it != adj_cells.end(); it++) {
+		if (top._prev_cell->getEdge(*it).canTraverse(mobility) && !(*it)->getTile().hasUnit()) { //if we can step on this tile
+			int cost = top._cost + top._prev_cell->getEdge(*it).getCost(mobility);
+			if (_shortest_path_map.find(*it) == _shortest_path_map.end()) {//no shortest path found
+				if (cost <= max_move || max_move == -1) {
+					prev_cost_pair new_element_insert;
+					prev_cost_pair new_element_push;
```

Comment:
It would be way less confusing if you were constructing all these instead of using the default constructor, then modifying them.
`
prev_cost_pair new_element_push{ *it, cost }
prev_cost_pair new_element_insert{ top._prev_cell, new_element_push._cost }
`